public/app/plugins/datasource/dashboard/datasource.ts

```diff
@@ -111,15 +125,162 @@ export class DashboardDatasource extends DataSourceApi<DashboardQuery> {
+    if (filterFieldIndices.length === 0) {
+      return frame;
+    }
+
+    // Short-circuit: if any filter has '=' operator with missing field, reject all rows
+    const hasImpossibleFilter = filterFieldIndices.some(({ fieldIndex }) => fieldIndex === -1);
+    if (hasImpossibleFilter) {
+      return this.reconstructDataFrame(frame, new Set<number>());
+    }
+
+    const matchingRows = new Set<number>();
+
+    // Check each row to see if it matches all filters (AND logic)
+    for (let rowIndex = 0; rowIndex < frame.length; rowIndex++) {
+      const rowMatches = filterFieldIndices.every(({ matcher, fieldIndex }) => {
+        // Handle case where field doesn't exist (fieldIndex === -1)
+        if (fieldIndex === -1) {
+          return false; // Impossible filter case
+        }
+
+        const field = frame.fields[fieldIndex];
+
+        // Use Grafana's value matcher system
+        return matcher!(rowIndex, field, frame, [frame]);
+      });
+
+      if (rowMatches) {
+        matchingRows.add(rowIndex);
+      }
     }
+
+    return this.reconstructDataFrame(frame, matchingRows);
+  }
+
+  /**
+   * Create a value matcher from an AdHoc filter
+   */
+  private createValueMatcher(filter: AdHocVariableFilter, fieldIndex: number, frame: DataFrame) {
+    // Return null for missing fields - they are handled separately
+    if (fieldIndex === -1) {
+      return null;
+    }
+
+    const field = frame.fields[fieldIndex];
+
+    // Only support string and numeric fields when feature toggle is enabled
+    if (config.featureToggles.dashboardDsAdHocFiltering) {
+      if (field.type !== FieldType.string && field.type !== FieldType.number) {
+        return null;
+      }
+    }
+
+    // Convert AdHoc filter to value matcher configuration
+    const matcherConfig = this.adHocFilterToMatcherConfig(filter);
+    if (!matcherConfig) {
+      return null;
+    }
+
+    try {
+      return getValueMatcher(matcherConfig);
+    } catch (error) {
+      console.warn('Failed to create value matcher for filter:', filter, error);
+      return null;
+    }
+  }
+
+  /**
+   * Convert AdHocVariableFilter to MatcherConfig for value matchers
+   */
+  private adHocFilterToMatcherConfig(filter: AdHocVariableFilter): MatcherConfig | null {
+    switch (filter.operator) {
+      case '=':
+        return {
+          id: ValueMatcherID.equal,
+          options: { value: filter.value },
+        };
+      case '!=':
+        return {
+          id: ValueMatcherID.notEqual,
+          options: { value: filter.value },
+        };
+      default:
+        // Unknown operator
+        return null;
+    }
+  }
+
+  /**
+   * Reconstruct DataFrame with only matching rows
+   * Optimized to avoid repeated array operations
+   */
+  private reconstructDataFrame(frame: DataFrame, matchingRows: Set<number>): DataFrame {
+    const fields: Field[] = frame.fields.map((field) => {
+      const newValues = Array.from(matchingRows, (rowIndex) => field.values[rowIndex]);
```

Comment:
Array.from() with mapper function iterates the Set twice (once to determine size, once to map values), which could be _super_ expensive if we have large data sets. 

If we do `new Array(matchingRows.size)`, we can pre-allocate the size for direct iteration, which should be much faster!

```suggestion
            // Pre-allocate array and iterate Set only once
      const newValues = new Array(matchingRows.size);
      let i = 0;
      for (const rowIndex of matchingRows) {
        newValues[i++] = field.values[rowIndex];
      }

```

Reply:
Thanks for this. Will make the change 👍 
I asked Claude, and learned a few things, I thought you might like to learn too.

---

Looking at this performance optimization suggestion, let me break down what's happening and analyze the claim:

## Current Code Analysis
```typescript
const newValues = Array.from(matchingRows, (rowIndex) => field.values[rowIndex]);
```

## The Optimization Claim
The comment suggests that `Array.from()` with a mapper function "iterates the Set twice" - but this isn't entirely accurate for modern JavaScript engines. Here's what actually happens:

### `Array.from()` behavior:
1. **Size determination**: Modern JS engines can get the size from `matchingRows.size` property directly (O(1) operation)
2. **Single iteration**: Creates and populates the array in one pass through the Set
3. **Internal array resizing**: Depending on the engine, there might be some internal array growth operations

### Suggested optimization benefits:
1. **Pre-allocation**: `new Array(matchingRows.size)` explicitly allocates the exact needed size upfront
2. **Direct iteration**: Manual loop avoids the callback overhead of `Array.from()`
3. **Memory efficiency**: Eliminates any potential array resizing during population

## Performance Impact

**For small datasets**: The difference would be negligible.

**For large datasets** (which seems to be the concern here): The optimization could provide measurable benefits because:
- Pre-allocation avoids array resizing operations
- Direct assignment is faster than callback invocation
- Better memory access patterns

## My Assessment

The optimization is **technically sound** and would provide a small performance improvement, especially for large datasets. However, the original claim about "iterating twice" is misleading - it's more about avoiding callback overhead and ensuring optimal memory allocation.

## Recommendation

Given that this is filtering dashboard data (which could indeed be large), and the code change is minimal while maintaining readability, **I'd recommend implementing the suggestion**. The performance gain may be small, but it's a worthwhile optimization in a hot path like data filtering.

The suggested code is also more explicit about its intent, which can be valuable for future maintainers.

---

And I asked when if ever we'd go with the original version of the code. Cursor / Claude said:
- **Manual loop is better when Dataset size > ~10,000 items**

We've got to expect that many items or more, so I think it's a great change 👍